app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java

```diff
@@ -0,0 +1,165 @@
+    private static final UiAutomation UI_AUTOMATION =
+        CustomUiDevice.getInstance().getInstrumentation().getUiAutomation();
+    private final List<MjpegScreenshotClient> clients;
+    private boolean isStopped = false;
+
+    MjpegScreenshotStream(List<MjpegScreenshotClient> clients) {
+        this.clients = clients;
+    }
+
+    @Override
+    public void interrupt() {
+        this.isStopped = true;
+        super.interrupt();
+    }
+
+    @Override
+    public void run() {
+        while (!isStopped) {
+            boolean noClientsConnected;
+            synchronized (clients) {
+                noClientsConnected = clients.isEmpty();
+            }
+
+            if (noClientsConnected) {
+                SystemClock.sleep(NO_CLIENTS_CONNECTED_SLEEP_TIME_MS);
+                continue;
+            }
+
+            // how long each loop should take in milliseconds
+            long targetInterval =
+                Math.round((1.0f / ServerConfig.getMjpegServerFramerate()) * 1000.0f);
+            long start = System.currentTimeMillis();
+
+            // add some resilience to handle `getScreenshot` or any of its
+            // sub-routines raising an error. If an error occurs during
+            // [post-]processing the frame will be dropped
+            byte[] screenshotData;
+            try {
+                screenshotData = getScreenshot();
+            } catch (Exception e) {
+                Logger.error("Error getting screenshot: ", e);
+                screenshotData = new byte[0];
+            }
+
+            if (screenshotData.length > 0) {
+                synchronized (clients) {
+                    Iterator<MjpegScreenshotClient> clientsIterator = clients.iterator();
+                    while (clientsIterator.hasNext()) {
+                        MjpegScreenshotClient client = clientsIterator.next();
+                        if (client.isClosed()) {
+                            clientsIterator.remove();
+                            continue;
+                        }
+
+                        if (!client.isInitialized()) {
+                            client.initialize();
+                        }
+
+                        client.write(screenshotData);
+                    }
+                }
+            }
+
+            // Always attempt to match the target framerate, even if the frame
+            // was dropped, to keep the rate consistent
+            matchFramerate(targetInterval, start);
+        }
+    }
+
+    private void matchFramerate(long targetInterval, long start) {
+        long end = System.currentTimeMillis();
+        long duration = end - start;
+
+        if (duration < targetInterval) {
+            SystemClock.sleep(targetInterval - duration);
+        }
+    }
+
+    private byte[] getScreenshot() {
+        Bitmap screenshot = UI_AUTOMATION.takeScreenshot();
+        if (screenshot == null) {
+            return new byte[0];
```

Comment:
how about throwing an exception instead, so we know why there where no screenshot?

Reply:
Sure, that's fine.

Reply:
This will be resolved w/ the upcoming update(s)